### PR TITLE
io: remove unsafe pin-projections and remove manual Unpin implementations

### PR DESCRIPTION
**Pull request by taiki-e**
Based on #1586 (First 2 commits are the same as #1586)

This removes most pin-projection related `unsafe` code: https://github.com/tokio-rs/tokio/commit/16b06e3bc1405f64b6dfb38aab2589fcabd579f7

And removes manual `Unpin` implementations and add tests to check that `Unpin` requirement does not change accidentally. see commit message for more: https://github.com/tokio-rs/tokio/commit/7ad962db620ea896755fec9d93fe3a68c02b0159

### CODE REVIEW

**Comment by carllerche** [comment]
Thanks for this. Do we know how much this impacts compile time?

**Comment by taiki-e** [comment]
I didn’t actually time it, but the master branch already depends on pin-project, so I think merging this should have little impact on compile time.

Impact of adding pin-project to dependencies: As pin-project is a proc-macro, it increases the compilation time. If the dependencies have other proc-macro crates, the impact will not be so big. (As far as I know, whether or not the "full" feature of `syn` is enabled will particularly affect compilation time. In fact, [a crate has been created to handle this issue](https://github.com/dtolnay/syn/issues/467). pin-project enables "full" feature.)

**Comment by taiki-e** [comment]
@carllerche I found the useful tool `-Z timings`: https://internals.rust-lang.org/t/exploring-crate-graph-build-times-with-cargo-build-ztimings/10975

Build times:

master:
master

this pr:
pr

without pin-project (613fde2)
613fde2

Dependencies are compiled in parallel, so I don't think they have much impact if another proc-macro crate (tokio-macros, tracing-attribute, etc.) is already enabled.

File: tokio-io/src/io/take.rs [hunk 1]
```diff
@@ -1,21 +1,21 @@
 use crate::{AsyncBufRead, AsyncRead};
 use futures_core::ready;
-use pin_utils::{unsafe_pinned, unsafe_unpinned};
+use pin_project::{pin_project, project};
 use std::pin::Pin;
 use std::task::{Context, Poll};
 use std::{cmp, io};
 
-/// Stream for the [`take`](super::AsyncReadExt::take) method.
+/// Reader for the [`take`](super::AsyncReadExt::take) method.
```
Comment (jonhoo) [inline tokio-io/src/io/take.rs]: Was this naming change intentional?

Comment (taiki-e) [reply tokio-io/src/io/take.rs]: Oops, I didn't intend to include that change in this PR.

File: tokio-io/src/io/chain.rs [hunk 2]
```diff
@@ -133,10 +125,16 @@ where
...
+            me.first.consume(amt)
         } else {
-            self.second().consume(amt)
+            me.second.consume(amt)
         }
     }
 }
+
+#[test]
+fn assert_unpin() {
```
Comment (taiki-e) [inline tokio-io/src/io/chain.rs]: Oh, I forgot to surround this with a `#[cfg(test)] mod tests {}` module. I will push a commit for this. (but, in fact, it seems to work properly without that module...)

**Comment by taiki-e** [comment]
FWIW: I have written a lightweight version of pin-project written with declarative macros: https://github.com/taiki-e/pin-project-lite

**Comment by taiki-e** [comment]
@carllerche  I used [`-Z timings` flag of cargo](https://internals.rust-lang.org/t/exploring-crate-graph-build-times-with-cargo-build-ztimings/10975): `cargo build -Z timings`

**Comment by carllerche** [comment]
Oh, nice, i missed all that :+1:

**Pull request merged**